tools/nodejs/src/statement.cpp

```diff
@@ -349,6 +353,62 @@ struct RunPreparedTask : public Task {
 
 			cb.MakeCallback(statement.Value(), {env.Null(), result_arr});
 		} break;
+		case RunType::ARROW_ALL: {
+			auto materialized_result = (duckdb::MaterializedQueryResult *)result.get();
+			// +1 is for null bytes at end of stream
+			Napi::Array result_arr(Napi::Array::New(env, materialized_result->RowCount() + 1));
+
+			auto deleter = [](Napi::Env, void *finalizeData, void *hint) {
+				delete static_cast<std::shared_ptr<duckdb::QueryResult> *>(hint);
+			};
+
+			std::shared_ptr<duckdb::QueryResult> result_ptr = move(result);
+
+			duckdb::idx_t out_idx = 1;
+			while (true) {
+				auto chunk = result_ptr->Fetch();
+
+				if (!chunk || chunk->size() == 0) {
+					break;
+				}
+
+				D_ASSERT(chunk->ColumnCount() == 2);
+				D_ASSERT(chunk->data[0].GetType() == duckdb::LogicalType::BLOB);
+				D_ASSERT(chunk->data[1].GetType() == duckdb::LogicalType::BOOLEAN);
+
+				for (duckdb::idx_t row_idx = 0; row_idx < chunk->size(); row_idx++) {
+					duckdb::string_t blob = ((duckdb::string_t *)(chunk->data[0].GetData()))[row_idx];
+					bool is_header = chunk->data[1].GetData()[row_idx];
+
+					// Create shared pointer to give (shared) ownership to ArrayBuffer, not that for these materialized
+					// query results, the string data is owned by the QueryResult
+					auto result_ref_ptr = new std::shared_ptr<duckdb::QueryResult>(result_ptr);
```

Comment:
Not for this PR but for the future - DuckDB creates a copy of strings when constructing the result object. Rather than keeping a reference to the result object, we can overwrite the result object creation for the node client using a [custom result collector](https://github.com/duckdb/duckdb/blob/master/src/include/duckdb/execution/operator/helper/physical_result_collector.hpp). Then we can directly create strings in a more compatible format (e.g. as `unique_ptr<data[]>`) so we don't need to keep a reference to the entire query result.

Reply:
ah interesting, yea that seems much nicer!